### ISSUE TITLE
Remote Assistance (with file invitation.msrcIncident) over proxy

### ISSUE THREAD

**Issue by boom1** [FreeRDP/FreeRDP#5712]
**Remote Assistance by invitation file does not work via proxy.**

Tested with xfreerdp as RDP client on Fedora 30 with FreeRDP from git repo, master branch of Nov. 7th.
Target system for assistance was Windows 10.

For testing, start Remote Assistance under Windows 10, with invitation by file :
> msra.exe /saveasfile invitation test-password

Copy the resulting file invitation.msrcincident to a Linux PC which can reach the Windows PC only via a proxy (in below example a SOCKS5 proxy at proxy.test.org, port 1080).

Start xfreerdp on the Linux PC, with proxy settings:
> xfreerdp invitation.msrcincident /proxy:socks5://proxy.test.org:1080 /assistance:test-password

This does not work (at least in my case it does not use the proxy).
And this patch should fix this: [FreeRDP.patch.txt](https://github.com/FreeRDP/FreeRDP/files/3831388/FreeRDP.patch.txt)

Basically the patch moves code that tries the multiple destinations given in an invitation file (AFAIK at least IPv4, IPv6 and DNS name of the Windows PC) from a lower layer to a higher layer. This makes proxies work, because they are handled in an intermediate layer which now is used for each destination from the invitation file that is tried.

"higher layer" is nego_tcp_connect() in nego.c
"intermediate layer" is transport_connect() in transport.c
"lower layer" is freerdp_tcp_connect_multi() in tcp.c

**Comment by akallabeth** [comment]
@boom1 so, does `proxy_prepare` not return `true` in your case?

**Comment by boom1** [comment]
From my memory: proxy_prepare() does return true and returns the proxy settings from the command line (because I have no environment variables for proxy set).

**Comment by akallabeth** [comment]
@boom1 checked it, looks like the glue is not correct for your use case.
The part with connecting to the proxy is lacking the alternative addresses.
the patch you provided is a nice start, but will fail for other cases (everything non NLA).
Mark it as todo.

**Comment by boom1** [comment]
I agree, the patch still fails for other cases (I did not test it, but gateways seems not to go through the proxy as well).
Thank you for marking it as todo.

**Comment by akallabeth** [comment]
@boom1 did some fixes in https://github.com/akallabeth/FreeRDP/tree/proxy_cleanup
does that start working for you?

**Comment by boom1** [comment]
I have Fedora 30 with openssl 1.1 and can't compile this version (which seems to rely on OpenSSL 1.0 includes).
It fails at winpr/libwinpr/crypto/crypto.h, lines 44/45 (incomplete type of EVP_CIPHER_CTX enc and EVP_CIPHER_CTX dec).
If I change them to be pointers, it goes a bit further but stops at winpr/libwinpr/sspi/NTLM/ntlm.c:887:11

**Comment by akallabeth** [comment]
@boom1 start from a clean checkout (and the correct branch). Theseare patches on master.

**Comment by boom1** [comment]
Sorry, I still had the master branch, but switched to proxy_cleanup now.

I get an segfault, but that seems to be an independent issue, which I had also in master with my patch. I solved it by disabling the free() in stream.c:129.

```
[...]
10:26:58:232] [20973:20979] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[10:26:58:232] [20973:20979] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[10:26:58:240] [20973:20979] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
[New Thread 0x7fffdd38b700 (LWP 20984)]
[New Thread 0x7fffdcb8a700 (LWP 20985)]
[New Thread 0x7fffd7fff700 (LWP 20986)]
[New Thread 0x7fffd77fe700 (LWP 20987)]
[10:26:58:255] [20973:20987] [INFO][com.freerdp.channels.rdpsnd.client] - Loaded fake backend for rdpsnd
[New Thread 0x7fffd6ffd700 (LWP 20988)]

Thread 2 "xfreerdp" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe532c700 (LWP 20979)]
0x00007ffff7ba21d0 in free () from /lib64/libc.so.6
[...removed complaint about missing debug info...]
(gdb) bt
#0  0x00007ffff7ba21d0 in free () from /lib64/libc.so.6
#1  0x00007ffff7d68ddf in Stream_Free (s=0x7fffe0000f10, bFreeBuffer=1) at winpr/libwinpr/utils/stream.c:129
#2  0x00007ffff7ea66ce in remdesk_virtual_channel_open_event_ex (lpUserParam=0x7fffe0000f10, openHandle=4, event=11, pData=0x7fffc8000b40, dataLength=58, totalLength=58, dataFlags=0) at channels/remdesk/client/remdesk_main.c:796
#3  0x00007ffff783ebc5 in freerdp_channels_process_message_free (message=0x7fffe532b3d0, type=11) at libfreerdp/core/client.c:471
#4  0x00007ffff783ec94 in freerdp_channels_process_message (instance=0x442ec0, message=0x7fffe532b3d0) at libfreerdp/core/client.c:504
#5  0x00007ffff783ecf0 in freerdp_channels_process_sync (channels=0x476c80, instance=0x442ec0) at libfreerdp/core/client.c:521
#6  0x00007ffff783eec2 in freerdp_channels_check_fds (channels=0x476c80, instance=0x442ec0) at libfreerdp/core/client.c:586
#7  0x00007ffff783b9ac in freerdp_check_event_handles (context=0x443150) at libfreerdp/core/freerdp.c:393
#8  0x00007ffff7f3df98 in xf_client_thread (param=0x442ec0) at client/X11/xf_client.c:1598
#9  0x00007ffff7d9b334 in thread_launcher (arg=0x4b1df0) at winpr/libwinpr/thread/thread.c:327
#10 0x00007ffff76334c0 in start_thread () from /lib64/libpthread.so.0
#11 0x00007ffff7c17553 in clone () from /lib64/libc.so.6
(gdb) 

```

**Comment by boom1** [comment]
To avoid the segfault I use below hack.
The proxy then works fine, thanks!

```
diff --git a/channels/remdesk/client/remdesk_main.c b/channels/remdesk/client/remdesk_main.c
index c950d458f..4431d42df 100644
--- a/channels/remdesk/client/remdesk_main.c
+++ b/channels/remdesk/client/remdesk_main.c
@@ -793,7 +793,7 @@ static VOID VCAPITYPE remdesk_virtual_channel_open_event_ex(LPVOID lpUserParam,
                case CHANNEL_EVENT_WRITE_COMPLETE:
                {
                        wStream* s = (wStream*)lpUserParam;
-                       Stream_Free(s, TRUE);
+//                     Stream_Free(s, TRUE);
                }
                break;

```

**Comment by boom1** [comment]
By the way, is there a  command line option to immediately request control over the remote assistance session (like pressing Ctrl-Alt-C)?

**Comment by akallabeth** [comment]
@boom1 ok, good to know the proxy now works.
the stream free fix you propose is a huge memory leak ;)

**Comment by boom1** [comment]
Yes, it's a memory leak. The patch was just meant as a quick hack to be able to continue working.
Am I the only one where this Stream_Free() crashes?
I had hoped that would have been solved in the meantime. I don't see any obvious errors, but I also don't really understand the event code (especially not where the message is created).
lpUserParam simply seems not to be a Stream sometimes.

**Comment by akallabeth** [comment]
@boom1 fix is in #5718

**Comment by akallabeth** [comment]
@boom1 better report such issues immediately (if you find them with nightlies / on a build from current master) as the library can not be tested automatically for every situation (there is some user interaction required, so this is only done case by case)

**Comment by boom1** [comment]
@akallabeth I was not so confident that this segfault wasn't my fault (because I did local changes and never had touched this project before).

**Comment by akallabeth** [comment]
@boom1 no worries, that one was mentioned by you first.
Most rely on downstream packages and there such stuff is usually recognized and fixed.
